mapeditor/mainwindow.cpp

```diff
@@ -429,6 +446,87 @@ void MainWindow::on_actionOpen_triggered()
 	openMap(filenameSelect);
 }
 
+void MainWindow::on_actionOpenRecent_triggered()
+{
+	QSettings s(Ui::teamName, Ui::appName);
+	QStringList recentFiles = s.value(recentlyOpenedFilesSetting).toStringList();
+
+	struct RecentFileDialog : public QDialog {
+		
+		RecentFileDialog(QWidget *parent, const QStringList& recentFiles) : QDialog(parent), layout(this) {
+
+			setWindowTitle("Recently Opened Files");
+			setMinimumWidth(600);
+
+			for (const QString &file : recentFiles) {
+				QListWidgetItem *item = new QListWidgetItem(file);
+				listWidget.addItem(item);
+			}
+
+			// Select most recent items by default.
+			// This enables a "CTRL+R => Enter"-workflow instead of "CTRL+R => 'mouse click on first item'"
+			if(listWidget.count() > 0) {
+				listWidget.item(0)->setSelected(true);
+			}
+
+			layout.setSizeConstraint(QLayout::SetMaximumSize);
+			layout.addWidget(&listWidget);
+		}
+
+		QVBoxLayout layout;
+		QListWidget listWidget;
+	};
+
+	RecentFileDialog d(this, recentFiles);
+
+	auto onSelect = [this, &d](QListWidgetItem *item) {
+		if(getAnswerAboutUnsavedChanges()) {
+			QString filename = item->text();
+			openMap(filename);
+		}
+		d.close();
+	};
+
+	connect(&d.listWidget, &QListWidget::itemActivated, this, onSelect);
```

Comment:
rather expose a signal from RecentFileDialog that sends file path, no need to leak internal implementation outside

Reply:
actually, a cleaner way would be to store selection internally and exposing it via a "getter", closing the dialog on its own. And the caller would use it like:
```qt
RecentFileDialog d{recentFiles, this};
if (d.exec())
  openMap(d.selectedFilePath());
```

Reply:
Good suggestions! Fixed with [5747781](https://github.com/vcmi/vcmi/pull/4939/commits/5747781dda52d97f17765603b96c15ab5597e350).